connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java

```diff
@@ -0,0 +1,220 @@
+        reporters.add(new JmxReporter(JMX_PREFIX));
+        this.metrics = new Metrics(metricConfig, reporters, time);
+        log.debug("Registering Connect metrics with JMX for worker '{}'", workerId);
+        AppInfoParser.registerAppInfo(JMX_PREFIX, workerId);
+    }
+
+    /**
+     * Get the worker identifier.
+     * @return the worker ID; never null
+     */
+    public String workerId() {
+        return workerId;
+    }
+
+    /**
+     * Get the {@link Metrics Kafka Metrics} that are managed by this object and that should be used to
+     * add sensors and individual metrics.
+     *
+     * @return the Kafka Metrics instance; never null
+     */
+    public Metrics metrics() {
+        return metrics;
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name.
+     *
+     * @param groupName the name of the metric group; may not be null
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName) {
+        return group(groupName, false);
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name and the given tags.
+     *
+     * @param groupName    the name of the metric group; may not be null
+     * @param tagKeyValues pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, String... tagKeyValues) {
+        return group(groupName, false, tagKeyValues);
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name and the given tags.
+     *
+     * @param groupName       the name of the metric group; may not be null
+     * @param includeWorkerId true if the tags should include the worker ID
+     * @param tagKeyValues    pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
+        MetricGroup group = groupsByName.get(groupName);
+        if (group == null) {
+            Map<String, String> tags = tags(includeWorkerId ? workerId : null, tagKeyValues);
+            group = new MetricGroup(groupName, tags);
+            groupsByName.put(groupName, group);
+        }
+        return group;
+    }
+
+    /**
+     * Create a set of tags using the supplied key and value pairs.
+     *
+     * @param workerId the worker ID that should be included first in the tags; may be null if not to be included
+     * @param keyValue the key and value pairs for the tags; must be an even number
+     * @return the map of tags that can be supplied to the {@link Metrics} methods; never null
+     */
+    static Map<String, String> tags(String workerId, String... keyValue) {
+        if ((keyValue.length % 2) != 0)
```

Comment:
Yeah, this is a case where I think it does becomes harder to read, especially because as you say it will be optimized.

Reply:
Yeah, there's some didactic aspect to a few lines that are just a bit harder to read of course. (for instance if it was `var` instead of `2` things would be different). But I was on the edge too. Fine with leaving it.